src/sentry/search/django/backend.py

```diff
@@ -139,12 +139,37 @@ def _build_queryset(self, project, query=None, status=None, tags=None,
             )
 
         if tags:
-            matches = self._tags_to_filter(project, tags)
-            if not matches:
-                return queryset.none()
-            queryset = queryset.filter(
-                id__in=matches,
-            )
+            filter_tags, exclude_tags = {}, {}
+            for k, v in six.iteritems(tags):
+                if k[0] == '-':
+                    exclude_tags[k[1:]] = v
+                else:
+                    filter_tags[k] = v
+            # Don't allow the same tag value to be both excluded and included
+            for k, v in six.iteritems(filter_tags):
+                if k in exclude_tags:
+                    if exclude_tags[k] == v:
+                        return queryset.none()
+                    else:
+                        # This exclude is unnecessary
+                        del exclude_tags[k]
+            if filter_tags:
+                matches = self._tags_to_filter(project, filter_tags)
+                if not matches:
+                    return queryset.none()
+                queryset = queryset.filter(
+                    id__in=matches,
+                )
+            if exclude_tags:
+                all_matches = []
+                for k, v in six.iteritems(exclude_tags):
+                    matches = self._tags_to_filter(project, {k: v})
+                    if not matches:
+                        continue
+                    all_matches += matches
+                queryset = queryset.exclude(
+                    id__in=set(all_matches),
```

Comment:
- It might make sense to do this set difference in memory between inclusion and exclusion tags rather than deferring to the database.
- It may also make sense that if we are dealing with inclusion tags to provide those IDs to the `_tags_to_filter` method here so that we don't get back results for groups that we were not part of the result set in the first place.